waveform_editor/configuration.py

```diff
@@ -115,9 +138,24 @@ def replace_waveform(self, waveform):
             raise ValueError(
                 f"Waveform '{waveform.name}' does not exist in the configuration."
             )
-        else:
-            group = self.waveform_map[waveform.name]
-            group.waveforms[waveform.name] = waveform
+
+        group = self.waveform_map[waveform.name]
+        old_waveform = group.waveforms[waveform.name]
+
+        group.waveforms[waveform.name] = waveform
+
+        try:
+            if isinstance(waveform, DerivedWaveform):
+                self.dependency_graph.add_node(
+                    waveform.name, waveform.dependent_waveforms
+                )
+            else:
+                self.dependency_graph.remove_node(waveform.name)
+            self._calculate_bounds()
+        except Exception as e:
+            # Revert replacement
+            group.waveforms[waveform.name] = old_waveform
```

Comment:
This risks desynchronizing your dependency_graph (which will have removed `waveform.name` when add_node failed).